### ISSUE TITLE
Pre-upload filters

### ISSUE THREAD

**Issue by DanielHeath** [camlistore/camlistore#847]
I'm currently using a mix of the android app, web UI and camput (all great).

However, portable devices tend not to have the required processing power for e.g. high-compression video encoding.

I have a script to convert photos to webp and videos to h.264 before uploading to camlistore (saves ~30% on space use), but there's no straightforward way to run it over android/web uploads.

**Comment by mpl** [comment]
@DanielHeath so you're saying you'd like a way for the android app to take into account these filters and apply them before uploading? I'm not sure that the role of the app to do that, but I understand the convenience, and I suppose you could argue that other "social apps" let you apply filters to your photos as well. But is it wise and doable to let the user provide their own?

As for the web uploads, I'd argue that if you're on a computer it's easy to apply the filters yourself. And if you're on mobile, hopefully one day you won't need to use to use the web UI as the app would provide all the (browsing) capabilities that the web UI. But that day may be pretty far ahead indeed, so I dunno.

**Comment by jmunson** [comment]
I have to agree with @mpl here -- while this would certainly be a convenient feature to have, I'd much rather have a long term solution where you can just upload your originals, then do any filtering or re-encoding either on the server side as a process you can apply manually or automatically,  or even on the server on a per-request basis the way e.g plex can take your movie files and as a client you can ask plex to transcode it to 480p while on a low bandwidth mobile connection, or 720p on a tablet at home, or directly stream the original bits to your desktop.

**Comment by DanielHeath** [comment]
I'd expect this would happen server-side. For this specific use-case (saving storage space) it would only work as a filter applied before uploaded files are split into blobs.

**Comment by mpl** [comment]
@DanielHeath What's your concrete proposal? How would you integrate these filters? in which part of Camlistore?

**Comment by DanielHeath** [comment]
So, I've had a more detailed look at the camlistore architecture and I don't think there's a way to do it that fits. 

Clients are expected to break files into appropriate sized blobs before sending, so any transcoding has to happen client-side. 

The reason the files are large in the first place is that my phone lacks the processing power to do a good/fast job of compression.

The only way I can see that might work is adding a custom upload endpoint which lets clients send a complete file and have the server respond with the ref it has created.

**Comment by tgulacsi** [comment]
Or do the transcoding "offline", i.e. after the client has uploaded the big video, transcode it in a separate app, upload the minified version, and delete the original.
Would work flawlessly, if we had GC implemented...

**Comment by DanielHeath** [comment]
In that case the transcoder could write a 'delete' claim linking to the new version which would be nice.

**Comment by kushtrimjunuzi** [comment]
How about just the events, notifying that file is uploaded? This could be something like a web hook.
